### PR TITLE
Light Client - use the new consensus types

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

The point of this PR is to switch from using protobufs to using light client types defined in `consensus-types`.

**Which issues(s) does this PR fix?**

Part of #12991 

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [ ] I have made an appropriate entry to [CHANGELOG.md](https://github.com/prysmaticlabs/prysm/blob/develop/CHANGELOG.md). **_Not adding an entry because we already have one about adding consensus types for light clients, and this PR is just a continuation of this_**
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

### CODE REVIEW

File: consensus-types/light-client/update.go [hunk 1]
```diff
@@ -23,6 +23,8 @@ func NewWrappedUpdate(m proto.Message) (interfaces.LightClientUpdate, error) {
 		return NewWrappedUpdateCapella(t)
 	case *pb.LightClientUpdateDeneb:
 		return NewWrappedUpdateDeneb(t)
+	case *pb.LightClientUpdateElectra:
```
Comment (rkapka) [inline consensus-types/light-client/update.go:26]: This was missing

File: beacon-chain/db/kv/lightclient.go [hunk 2]
```diff
@@ -5,35 +5,41 @@ import (
...
-	ethpbv2 "github.com/prysmaticlabs/prysm/v5/proto/eth/v2"
+	ethpb "github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v5/runtime/version"
 	bolt "go.etcd.io/bbolt"
+	"google.golang.org/protobuf/proto"
 )
 
-func (s *Store) SaveLightClientUpdate(ctx context.Context, period uint64, update *ethpbv2.LightClientUpdateWithVersion) error {
-	ctx, span := trace.StartSpan(ctx, "BeaconDB.saveLightClientUpdate")
+func (s *Store) SaveLightClientUpdate(ctx context.Context, period uint64, update interfaces.LightClientUpdate) error {
```
Comment (rkapka) [inline beacon-chain/db/kv/lightclient.go:20]: We can save light client updates similarly how we save blocks - by prepending the version to encoded bytes. That way we don't need the separate `XXXWithVersion` type

File: beacon-chain/core/light-client/lightclient.go [hunk 3]
```diff
@@ -299,46 +329,36 @@ func ComputeWithdrawalsRoot(payload interfaces.ExecutionData) ([]byte, error) {
 	return withdrawalsRoot, nil
 }
 
-func BlockToLightClientHeader(block interfaces.ReadOnlySignedBeaconBlock) (*ethpbv2.LightClientHeaderContainer, error) {
-	switch block.Version() {
+func BlockToLightClientHeader(ver int, block interfaces.ReadOnlySignedBeaconBlock) (interfaces.LightClientHeader, error) {
```
Comment (rkapka) [inline beacon-chain/core/light-client/lightclient.go:332]: As discussed, I added a version parameter which is populated from the attested block's version in `NewLightClientUpdateFromBeaconState`

File: beacon-chain/blockchain/process_block_helpers.go [hunk 4]
```diff
@@ -182,23 +182,23 @@ func (s *Service) sendLightClientFinalityUpdate(ctx context.Context, signed inte
...
-
 	if err != nil {
 		return 0, errors.Wrap(err, "could not create light client update")
 	}
 
 	// Return the result
 	result := &ethpbv2.LightClientFinalityUpdateWithVersion{
 		Version: ethpbv2.Version(signed.Version()),
-		Data:    update,
+		// TODO: Get back to this when revisiting events
```
Comment (rkapka) [inline beacon-chain/blockchain/process_block_helpers.go:200]: I commented out everything event-related. We need to revisit events anyway, so dealing with this now is not that important.

File: beacon-chain/rpc/eth/light-client/helpers.go [hunk 5]
```diff
@@ -68,14 +71,13 @@ func createLightClientBootstrapAltair(ctx context.Context, state state.BeaconSta
 		return nil, fmt.Errorf("latest block header root %#x not equal to block root %#x", latestBlockHeaderRoot, beaconBlockRoot)
 	}
 
-	lightClientHeaderContainer, err := lightclient.BlockToLightClientHeader(block)
+	lightClientHeader, err := lightclient.BlockToLightClientHeader(block.Version(), block)
```
Comment (rkapka) [inline beacon-chain/rpc/eth/light-client/helpers.go:74]: Is sing the block's version fine?

File: beacon-chain/rpc/eth/light-client/helpers.go [hunk 6]
```diff
@@ -136,14 +141,13 @@ func createLightClientBootstrapCapella(ctx context.Context, state state.BeaconSt
...
```
Comment (rkapka) [inline beacon-chain/rpc/eth/light-client/helpers.go:144]: Is using the block's version fine?

File: beacon-chain/core/light-client/lightclient.go (hunk 3 above)
Comment (Inspector-Butters) [reply beacon-chain/core/light-client/lightclient.go:332]: Could this potentially cause a problem? for example when version is Capella but block is Altair. will calling the `block.execution` getters on an Altair block result in an error?

File: beacon-chain/rpc/eth/light-client/helpers.go (hunk 5 above)
Comment (Inspector-Butters) [reply beacon-chain/rpc/eth/light-client/helpers.go:74]: yes it should be.

**Review by rupam-04 — APPROVED** [review]
lgtm

File: beacon-chain/core/light-client/lightclient.go [hunk 7]
```diff
@@ -29,13 +28,14 @@ const (
...
 	ctx context.Context,
+	currentSlot primitives.Slot,
 	state state.BeaconState,
 	block interfaces.ReadOnlySignedBeaconBlock,
 	attestedState state.BeaconState,
 	attestedBlock interfaces.ReadOnlySignedBeaconBlock,
 	finalizedBlock interfaces.ReadOnlySignedBeaconBlock,
 ) (interfaces.LightClientFinalityUpdate, error) {
-	update, err := NewLightClientUpdateFromBeaconState(ctx, state, block, attestedState, attestedBlock, finalizedBlock)
+	update, err := NewLightClientUpdateFromBeaconState(ctx, currentSlot, state, block, attestedState, attestedBlock, finalizedBlock)
```
Comment (Inspector-Butters) [inline beacon-chain/core/light-client/lightclient.go:38]: Can we access the current slot using the `state` or `block` that is passed? they are the current state and block.

File: beacon-chain/core/light-client/lightclient.go [hunk 8]
```diff
@@ -194,7 +203,7 @@ func NewLightClientUpdateFromBeaconState(
 		// if finalized_block.message.slot != GENESIS_SLOT
 		if finalizedBlock.Block().Slot() != 0 {
 			// update.finalized_header = block_to_light_client_header(finalized_block)
-			finalizedLightClientHeader, err := BlockToLightClientHeader(attestedBlock.Version(), finalizedBlock)
+			finalizedLightClientHeader, err := BlockToLightClientHeader(ctx, currentSlot, finalizedBlock)
```
Comment (Inspector-Butters) [inline beacon-chain/core/light-client/lightclient.go:206]: I am still curious to see if anything bad happens when you call this function with a Capella slot on an Altair/Bellatrix block.

File: beacon-chain/rpc/endpoints.go [hunk 9]
```diff
@@ -848,10 +848,11 @@ func (*Service) configEndpoints() []endpoint {
...
 func (s *Service) lightClientEndpoints(blocker lookup.Blocker, stater lookup.Stater) []endpoint {
 	server := &lightclient.Server{
-		Blocker:     blocker,
-		Stater:      stater,
-		HeadFetcher: s.cfg.HeadFetcher,
-		BeaconDB:    s.cfg.BeaconDB,
+		Blocker:          blocker,
+		Stater:           stater,
+		HeadFetcher:      s.cfg.HeadFetcher,
+		ChainInfoFetcher: s.cfg.ChainInfoFetcher,
```
Comment (Inspector-Butters) [inline beacon-chain/rpc/endpoints.go:854]: doesn't the `HeadFetcher` gives us access to the current slot? `HeadFetcher.HeadState.Slot`

File: beacon-chain/core/light-client/lightclient.go (hunk 8 above)
Comment (rupam-04) [reply beacon-chain/core/light-client/lightclient.go:206]: > I am still curious to see if anything bad happens when you call this function with a Capella slot on an Altair/Bellatrix block.

Is that possible? A slot can only have a single block or no block at all, shouldn't their versions stay the same?

**Pull request merged**